lib/src/impl_ffi/impl_ffi.utils.dart

```diff
@@ -14,40 +14,101 @@
-/// automatically be called with [key] is garbage collected by Dart.
-///
-/// This takes ownership of [key], and the caller **may not** call
-/// [ssl.EVP_PKEY_free].
+/// Function that can't be inlined to be used for preventing [obj] from being
+/// garbage collected.
+@pragma('vm:never-inline')
+Object _finalizerReachabilityFence(Object obj) {
+  return obj;
+}
+
+/// Wrapper around [EVP_PKEY] which attaches finalizer and ensures that the
+/// [_finalizerReachabilityFence] is used after usage.
 ///
-/// Callers should be ware that the Dart GC may collect [key] as soon as it
-/// deems the object to not be in use anymore. This can happy at any point when
-/// the VM (optimizer) determines that no code-path passing [key] to an FFI
-/// function can be called again. For this reason, users should take extra care
-/// to make sure that all accesses to [key] takes an extra reference.
-void _attachFinalizerEVP_PKEY(ffi.Pointer<EVP_PKEY> key) {
-  final ret = dl.webcrypto_dart_dl_attach_finalizer(
-    key,
-    key.cast(),
-    ssl.addresses.EVP_PKEY_free.cast(),
-    // We don't really have an estimate of how much space the EVP_PKEY structure
-    // takes up, but if we make it some non-trivial size then hopefully the GC
-    // will prioritize freeing them.
-    4096,
-  );
-  if (ret != 1) {
-    throw AssertionError('package:webcrypto failed to attached finalizer');
+/// The [_finalizerReachabilityFence] ensures that the [_EvpPKey] is not garbage
+/// collected while the wrapped key is use. Thus, we can be sure the finalizer
+/// won't be called while the wrapped key is in use.
+class _EvpPKey {
+  final ffi.Pointer<EVP_PKEY> _pkey;
+
+  /// Allocate new [EVP_PKEY], attach finalizer and return the wrapped key.
+  factory _EvpPKey() {
+    final _pkey = ssl.EVP_PKEY_new();
+    _checkOp(_pkey.address != 0, fallback: 'allocation failure');
+    return _EvpPKey.wrap(_pkey);
+  }
+
+  /// Wrap existing [EVP_PKEY], this will attach a finalizer.
+  ///
+  /// After this, the wrapped key may only be used within a callback passed to
+  /// [use]. Otherwise, the garbage collect may be calling the finalizer while
+  /// the key is in use.
+  _EvpPKey.wrap(this._pkey) {
+    final ret = dl.webcrypto_dart_dl_attach_finalizer(
+      this,
+      _pkey.cast(),
+      ssl.addresses.EVP_PKEY_free.cast(),
+      // We don't really have an estimate of how much space the EVP_PKEY structure
+      // takes up, but if we make it some non-trivial size then hopefully the GC
+      // will prioritize freeing them.
+      4096,
+    );
+    if (ret != 1) {
+      throw AssertionError('package:webcrypto failed to attached finalizer');
+    }
+  }
+
+  /// Use the wrapped [EVP_PKEY] in callback [fn].
+  ///
+  /// Note. [fn] is not allowed to return a [Future].
+  T use<T>(T Function(ffi.Pointer<EVP_PKEY> pkey) fn) {
+    try {
+      return fn(_pkey);
+    } finally {
+      _finalizerReachabilityFence(this);
+    }
   }
```

Comment:
This is actually find for all use-cases (at-least as far I can understand), the BoringSSL C library has internal reference counting, so all the methods seem as though they take an `EVP_PKEY` and then they hold a reference to it while doing whatever they need to do.